include/exec/repeat_effect_until.hpp

```diff
@@ -0,0 +1,219 @@
+      friend void tag_invoke(set_value_t, _Self &&__self, _Done &&__done_ish) noexcept {
+        bool __done = static_cast<bool>(__done_ish); // BUGBUG potentially throwing.
+        auto *__op = __self.__op_;
+
+        // The following line causes the invalidation of __self.
+        __op->__source_op_.__destruct();
+
+        // If the sender completed with true, we're done
+        if (__done) {
+          stdexec::set_value((_Receiver &&) __op->__rcvr_);
+        } else {
+          try {
+            auto &__source_op = __op->__source_op_.__construct_with([&]() {
+              return stdexec::connect(stdexec::on(__op->__sched_, __op->__source_), __t{__op});
+            });
+            stdexec::start(__source_op);
+          } catch (...) {
+            stdexec::set_error((_Receiver &&) __op->__rcvr_, std::current_exception());
+          }
+        }
+      }
+
+      template <same_as<__t> _Self>
+        requires __callable<set_stopped_t, _Receiver>
+      friend void tag_invoke(set_stopped_t, _Self &&__self) noexcept {
+        auto *__op = __self.__op_;
+        __op->__source_op_.__destruct();
+        stdexec::set_stopped((_Receiver &&) __op->__rcvr_);
+      }
+
+      template <same_as<__t> _Self, class _Error>
+        requires __callable<set_error_t, _Receiver, _Error>
+      friend void tag_invoke(set_error_t, _Self &&__self, _Error __error) noexcept {
+        auto *__op = __self.__op_;
+        __op->__source_op_.__destruct();
+        stdexec::set_error((_Receiver &&) __op->__rcvr_, (_Error &&) __error);
+      }
+
+      friend env_of_t<_Receiver> tag_invoke(get_env_t, const __t &__self) noexcept(
+        __nothrow_callable<get_env_t, const _Receiver &>) {
+        return get_env(__self.__op_->__rcvr_);
+      }
+    };
+
+    template <class _SourceId>
+    struct __sender {
+      using _Source = stdexec::__t<_SourceId>;
+
+      template <class _Receiver>
+      using __op_t = stdexec::__t< __operation<_SourceId, stdexec::__id<_Receiver>>>;
+
+      template <class _Receiver>
+      using __receiver_t = stdexec::__t< __receiver<_SourceId, stdexec::__id<_Receiver>>>;
+
+      struct __t {
+        using __id = __sender;
+        [[no_unique_address]] _Source __source_;
+
+        template <class... Ts>
+        using __value_t = stdexec::completion_signatures<>;
+
+        template <class _Env>
+        using __completion_signatures = //
+          stdexec::make_completion_signatures<
+            _Source &,
+            _Env,
+            completion_signatures<set_error_t(std::exception_ptr), stdexec::set_value_t()>,
+            __value_t>;
+
+        template <__decays_to<__t> _Self, class _Env>
+        friend auto tag_invoke(get_completion_signatures_t, _Self &&, _Env)
+          -> __completion_signatures<_Env> {
+          return {};
+        }
+
+        template <class _Source2>
+        explicit __t(_Source2 &&__source) noexcept(__nothrow_decay_copyable<_Source2>)
```

Comment:
This shadows the compiler generated constructors.

Reply:
What I mean is to constrain `Source2` with something like `template <__decays_to<_Source> _Source2>`

Reply:
Yeah, I'm sorry, it looks like I missed the constraint here

Reply:
No need to be sorry! This is awesome work